File: fastapi_app/fastapi_app/services/prediction_service.py
from fastapi_app.data.market_data_client import MarketDataClient
from fastapi_app.ml.feature_engineering import build_dataset, features
from fastapi_app.ml.smile_model import SmileModel
from fastapi_app.models import Prediction, PredictionRequest, HistoryResponse
from datetime import datetime, timezone
from typing import Dict
import numpy as np

class PredictionService:
    def __init__(self, market_client: MarketDataClient, rationale_service=None):
        self.market = market_client
        self.rationale = rationale_service
        self.models: Dict[str, SmileModel] = {}

    def train(self, symbol: str) -> SmileModel:
        hist: HistoryResponse = self.market.get_daily_history(symbol, "3mo")
        X, y = build_dataset(hist.bars)
        m = SmileModel()
        m.fit(X, y)
        self.models[symbol] = m
        return m

    def predict(self, symbol: str, horizon: str) -> Prediction:
        hist: HistoryResponse = self.market.get_daily_history(symbol, "3mo")
        bars = hist.bars
        X, y = build_dataset(bars)
        model = self.models.get(symbol)
        if model is None:
            m = SmileModel()
            m.fit(X, y)
            model = m
            self.models[symbol] = model
        if not bars:
            return Prediction(symbol=symbol, horizon=horizon, as_of=datetime.now(timezone.utc), last_price=0.0, predicted_move="UP", confidence=0.5, rationale=None)
        f = features(bars, len(bars) - 1)
        p_up = model.predict_proba(f)
        move = "UP" if p_up >= 0.5 else "DOWN"
        last = bars[-1].close
        rationale = None
        if self.rationale:
            rationale = self.rationale.explain(symbol, horizon, p_up, move, bars)
        return Prediction(symbol=symbol, horizon=horizon, as_of=datetime.now(timezone.utc), last_price=last, predicted_move=move, confidence=abs(p_up - 0.5) * 2, rationale=rationale)

File: archive/java-stock-predictor/src/main/java/com/PredictStockPrice/stocks/service/PredictionService.java
package com.PredictStockPrice.stocks.service;

import java.time.Instant;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;

import org.springframework.cache.annotation.Cacheable;
import org.springframework.stereotype.Service;

import com.PredictStockPrice.stocks.data.MarketDataClient;
import com.PredictStockPrice.stocks.dto.HistoryResponse;
import com.PredictStockPrice.stocks.ml.FeatureEngineering;
import com.PredictStockPrice.stocks.ml.SmileModel;
import com.PredictStockPrice.stocks.model.Prediction;

import lombok.RequiredArgsConstructor;

@Service
@RequiredArgsConstructor
public class PredictionService {

    private final MarketDataClient marketDataClient;
    private final RationaleService rationaleService; // optional, can be no-op
    private final Map<String, SmileModel> models = new ConcurrentHashMap<>();

    @Cacheable("models")
    public SmileModel train(String symbol) {
        HistoryResponse hist = marketDataClient.getDailyHistory(symbol, "3mo");
        var ds = FeatureEngineering.buildDataset(hist.bars());
        SmileModel model = new SmileModel();
        model.fit(ds.X(), ds.y());
        models.put(symbol, model);
        return model;
    }

    public Prediction predict(String symbol, String horizon) {
        var hist = marketDataClient.getDailyHistory(symbol, "3mo");
        var bars = hist.bars();
        var ds = FeatureEngineering.buildDataset(bars);
        SmileModel model = models.computeIfAbsent(symbol, s -> {
            SmileModel m = new SmileModel();
            m.fit(ds.X(), ds.y());
            return m;
        });
        if (bars.isEmpty()) {
            return Prediction.builder()
                    .symbol(symbol).horizon(horizon).asOf(Instant.now())
                    .lastPrice(0).predictedMove(Prediction.Move.UP).confidence(0.5).build();
        }

        double[] f = FeatureEngineering.features(bars, bars.size() - 1);
        double pUp = model.predictProba(f);
        var move = pUp >= 0.5 ? Prediction.Move.UP : Prediction.Move.DOWN;
        var last = bars.get(bars.size() - 1).getClose();
        String rationale = rationaleService.explain(symbol, horizon, pUp, move, bars);

        return Prediction.builder()
                .symbol(symbol)
                .horizon(horizon)
                .asOf(Instant.now())
                .lastPrice(last)
                .predictedMove(move)
                .confidence(Math.abs(pUp - 0.5) * 2) // convert to 0..1 distance
                .rationale(rationale)
                .build();
    }
}
